cmd/snap/complete.go

```diff
@@ -120,46 +212,76 @@ func (cps connectPlugSpec) Complete(match string) []flags.Completion {
 
 	var ret []flags.Completion
 
-	var snapPrefix, plugPrefix string
+	var prefix string
 	if len(parts) == 2 {
 		// The user typed the colon, means they know the snap they want;
 		// go with that.
-		plugPrefix = parts[1]
+		prefix = parts[1]
 		snaps[parts[0]] = true
 	} else {
 		// The user is about to or has started typing a snap name but didn't
 		// reach the colon yet. Offer plugs for snaps with names that start
 		// like that.
-		snapPrefix = parts[0]
-		for _, plug := range ifaces.Plugs {
-			if strings.HasPrefix(plug.Snap, snapPrefix) {
-				snaps[plug.Snap] = true
+		snapPrefix := parts[0]
+		if spec.plugs {
+			for _, plug := range ifaces.Plugs {
+				if strings.HasPrefix(plug.Snap, snapPrefix) && spec.connFilter(len(plug.Connections)) {
+					snaps[plug.Snap] = true
+				}
+			}
+		}
+		if spec.slots {
+			for _, slot := range ifaces.Slots {
+				if strings.HasPrefix(slot.Snap, snapPrefix) && spec.connFilter(len(slot.Connections)) {
+					snaps[slot.Snap] = true
+				}
 			}
 		}
 	}
 
 	if len(snaps) == 1 {
 		for snapName := range snaps {
-			for _, plug := range ifaces.Plugs {
-				if plug.Snap == snapName && strings.HasPrefix(plug.Name, plugPrefix) {
-					// TODO: in the future annotate plugs that can take
-					// multiple connection sensibly and don't skip those even
-					// if they have connections already.
-					if len(plug.Connections) == 0 {
-						ret = append(ret, flags.Completion{Item: fmt.Sprintf("%s:%s", plug.Snap, plug.Name)})
+			actualName := snapName
+			if spec.plugs {
+				if spec.connected && snapName == "" {
+					actualName = "core"
+				}
+				for _, plug := range ifaces.Plugs {
+					if plug.Snap == actualName && strings.HasPrefix(plug.Name, prefix) && spec.connFilter(len(plug.Connections)) {
+						// TODO: in the future annotate plugs that can take
+						// multiple connection sensibly and don't skip those even
+						// if they have connections already.
+						ret = append(ret, flags.Completion{Item: fmt.Sprintf("%s:%s", snapName, plug.Name), Description: "plug"})
+					}
+				}
+			}
+			if spec.slots {
+				if actualName == "" {
```

Comment:
counter-nitpick: that'll result in setting it to core when it's already core, in some situations :-)